src/collectives.c

```diff
@@ -1047,3 +1048,116 @@ shmem_internal_fcollect_recdbl(void *target, const void *source, size_t len,
 
     shmem_internal_quiet();
 }
+
+
+/* Circulator iterator for PE active sets */
+static inline int
+shmem_internal_circular_iter_next(int curr, int PE_start, int logPE_stride, int PE_size)
+{
+    const int stride = 1 << logPE_stride;
+    const int last = PE_start + (stride * (PE_size - 1));
+    int next;
+
+    next = curr + stride;
+    if (next > last)
+        next = PE_start;
+
+    return next;
+}
+
+
+void
+shmem_internal_alltoall(void *dest, const void *source, size_t len,
+                        int PE_start, int logPE_stride, int PE_size, long *pSync)
+{
+    const long one = 1;
+    const int stride = 1 << logPE_stride;
+    const int my_as_rank = (shmem_internal_my_pe - PE_start) / stride;
+    const void *dest_ptr = (uint8_t *) dest + my_as_rank * len;
+    int peer;
+
+    if (0 == len)
+        return;
```

Comment:
Log an issue to check if length == 0 is handled correctly for the other collectives?